FrEee/Scripts/AI/Vanguard/MinistryOfInfrastructure.csx

```diff
@@ -0,0 +1,246 @@
+                var cmd = new AddOrderCommand
+                (
+                    colony.ConstructionQueue,
+                    order
+                );
+                Empire.Commands.Add(cmd);
+                Empire.LogAIMessage($"Construction of a {chosenFacility.Name} on {planet.Name} has been ordered");
+            }
+        }
+    }
+
+
+    /// <summary>
+    /// Ensure that all vital system options are taken care of. 
+    /// </summary>
+    /// <param name="starSystem"></param>
+    /// <returns></returns>
+    protected FacilityTemplate GetEmpireVitalFaciltiesForSystem(StarSystem starSystem)
+    {
+        //spaceports. 
+        var option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Spaceport"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Spaceport"))
+            return option;
+        //resupply. 
+        option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Supply Generation"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Supply Generation"))
+            return option;
+        //spaceyard. Make sure every system has at least one. 
+        //TODO; mod might make more than one facility be a space yard... make sure to select the right one. 
+        option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Space Yard"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Space Yard"))
+            return option;
+        //warp point opening. Lets not let the player drop a fleet into the capital system. 
+        option = AvailableFacilities.FirstOrDefault(x => x.HasAbility("Stop Open Warp Point"));
+        if (option != null && !HasStarSystemFacility(starSystem, "Stop Open Warp Point"))
+            return option;
+
+
+
+        return null; 
+    }
+
+    /// <summary>
+    /// Any facility that would be vital for the planet...
+    /// </summary>
+    /// <param name="planet"></param>
+    /// <returns></returns>
+    protected FacilityTemplate GetPlanetVitalFacilities(Planet planet)
+    {
+        //TODO: handle things like planet increasing planets. 
+        return null; 
+    }
+
+    /// <summary>
+    /// Chooses what resource facility to build on this planet. 
+    /// </summary>
+    /// <param name="planet"></param>
+    /// <returns></returns>
+    protected FacilityTemplate GetResourceFacility(Planet planet, int randMin = 10, int randMax = 100)
+    {
+        return (from faci in AvailableFacilities
+        where (faci.HasAbility("Resource Generation - Minerals")
+           || faci.HasAbility("Resource Generation - Organics")
+           || faci.HasAbility("Resource Generation - Radioactives")
+           || faci.HasAbility("Point Generation - Research")
+           || faci.HasAbility("Point Generation - Intelligence"))
+           //ensure a little random in the scores, so not always building what is best. 
+        orderby GetScoreForFacility(planet, faci) + RandomHelper.Range(randMin, randMax) descending
```

Comment:
I wonder if we should use a PRNG here for consistent pseudo-random results?

Reply:
Not sure. Consistant could mean predictable, which is bad on the strategic level.